File: _C/src/python/bindings/math.cc
#include "jetdl/math.h"

#include <pybind11/pybind11.h>
#include <pybind11/stl.h>

#include <cmath>
#include <memory>
#include <stdexcept>
#include <vector>

#include "jetdl/python/math/bindings.h"

namespace py = pybind11;

namespace jetdl {

void bind_math_submodule(py::module_& m) {
  py::module_ math = m.def_submodule("math");
  math.def("c_add", &math::add, py::call_guard<py::gil_scoped_release>());
  math.def("c_sub", &math::sub, py::call_guard<py::gil_scoped_release>());
  math.def("c_mul", &math::mul, py::call_guard<py::gil_scoped_release>());
  math.def("c_div", &math::div, py::call_guard<py::gil_scoped_release>());
  math.def("c_pow", &math::pow);
  math.def("c_scalar_sqrt", [](py::object& input) {
    if (py::isinstance<py::int_>(input) || py::isinstance<py::float_>(input)) {
      return std::sqrt(py::cast<float>(input));
    } else {
      throw std::runtime_error(
          "INTERNAL: c_scalar_sqrt outside int & float not supported.\n");
    }
  });
  math.def("c_sqrt", &math::sqrt);
  math.def("c_sum",
           [](std::shared_ptr<Tensor>& tensor, const py::object& axes) {
             if (py::isinstance<py::int_>(axes)) {
               auto axes_vec = std::vector<int>(1, py::cast<int>(axes));
               return math::sum(tensor, axes_vec);
             } else if (py::isinstance<py::list>(axes) ||
                        py::isinstance<py::tuple>(axes)) {
               auto axes_vec = axes.cast<std::vector<int>>();
               return math::sum(tensor, axes_vec);
             } else if (axes.is_none()) {
               auto axes_vec = std::vector<int>{};
               return math::sum(tensor, axes_vec);
             } else {
               throw std::runtime_error(
                   py::str("Invalid data type for axes. Got type '{}'")
                       .format(py::type::of(axes)));
             }
           });

  math.def("c_sum_to_shape", [](std::shared_ptr<Tensor>& self,
                                const py::object& shape) {
    if (py::isinstance<py::list>(shape) || py::isinstance<py::tuple>(shape)) {
      auto shape_vec = py::cast<std::vector<size_t>>(shape);
      return math::sum_to_shape(self, shape_vec);
    } else {
      throw py::type_error(
          py::str("type '{}' for shape input for tensor reduction.")
              .format(py::type::of(shape)));
    }
  });

  math.def("c_exp", &math::exp, py::call_guard<py::gil_scoped_release>());
  math.def("c_log", &math::log, py::call_guard<py::gil_scoped_release>());
  math.def("c_log10", &math::log10, py::call_guard<py::gil_scoped_release>());
  math.def("c_log2", &math::log2, py::call_guard<py::gil_scoped_release>());
  math.def("c_sin", &math::sin, py::call_guard<py::gil_scoped_release>());
  math.def("c_cos", &math::cos, py::call_guard<py::gil_scoped_release>());
  math.def("c_tanh", &math::tanh, py::call_guard<py::gil_scoped_release>());
  math.def("c_sinh", &math::sinh, py::call_guard<py::gil_scoped_release>());
  math.def("c_cosh", &math::cosh, py::call_guard<py::gil_scoped_release>());
  math.def("c_abs", &math::abs, py::call_guard<py::gil_scoped_release>());
  math.def("c_sign", &math::sign, py::call_guard<py::gil_scoped_release>());
  math.def("c_clamp", &math::clamp, py::call_guard<py::gil_scoped_release>());

  math.def("c_mean",
           [](std::shared_ptr<Tensor>& tensor, const py::object& axes) {
             if (py::isinstance<py::int_>(axes)) {
               auto axes_vec = std::vector<int>(1, py::cast<int>(axes));
               return math::mean(tensor, axes_vec);
             } else if (py::isinstance<py::list>(axes) ||
                        py::isinstance<py::tuple>(axes)) {
               auto axes_vec = axes.cast<std::vector<int>>();
               return math::mean(tensor, axes_vec);
             } else if (axes.is_none()) {
               auto axes_vec = std::vector<int>{};
               return math::mean(tensor, axes_vec);
             } else {
               throw std::runtime_error(
                   py::str("Invalid data type for axes. Got type '{}'")
                       .format(py::type::of(axes)));
             }
           });
}

}  // namespace jetdl

File: jetdl/math.py
from typing import Optional, Union

from ._C import Tensor
from ._C.math import (c_abs, c_add, c_clamp, c_cos, c_cosh, c_div, c_exp,
                      c_log, c_log10, c_log2, c_mean, c_mul, c_pow,
                      c_scalar_sqrt, c_sign, c_sin, c_sinh, c_sqrt, c_sub,
                      c_sum, c_sum_to_shape, c_tanh)


def add(a: Tensor, b: Tensor) -> Tensor:
    """Computes the element-wise addition of two tensors.

    Args:
        a (Tensor): The first tensor.
        b (Tensor): The second tensor.
    """
    return c_add(a, b)


def sub(a: Tensor, b: Tensor) -> Tensor:
    """Computes the element-wise subtraction of two tensors.

    Args:
        a (Tensor): The first tensor.
        b (Tensor): The second tensor.
    """
    return c_sub(a, b)


def mul(a: Tensor, b: Tensor) -> Tensor:
    """Computes the element-wise multiplication of two tensors.

    Args:
        a (Tensor): The first tensor.
        b (Tensor): The second tensor.
    """
    return c_mul(a, b)


def div(a: Tensor, b: Tensor) -> Tensor:
    """Computes the element-wise division of two tensors.

    Args:
        a (Tensor): The first tensor.
        b (Tensor): The second tensor.
    """
    return c_div(a, b)


def pow(input: Tensor, exponent: int) -> Tensor:
    """Computes the element-wise power of a tensor.

    Args:
        input (Tensor): The input tensor.
        exponent (int): The exponent.
    """
    return c_pow(input, exponent)


def sqrt(input: Union[int, float, Tensor]) -> Union[int, float, Tensor]:
    """Computes the element-wise square root of a tensor or a scalar.

    Args:
        input (Union[int, float, Tensor]): The input tensor or scalar.
    """
    if isinstance(input, (int, float)):
        return c_scalar_sqrt(input)
    else:
        return c_sqrt(input)


def sum(input: Tensor, axes: Optional[Union[int, list, tuple]] = None) -> Tensor:
    """Computes the sum of tensor elements over given axes.

    Args:
        input (Tensor): The input tensor.
        axes (Optional[Union[int, list, tuple]]): The axes to reduce.
    """
    return c_sum(input, axes)


def sum_to_shape(input: Tensor, shape: Union[list, tuple]) -> Tensor:
    """Sums the tensor to a desired shape.

    Args:
        input (Tensor): The input tensor.
        shape (Union[list, tuple]): The desired shape.
    """
    return c_sum_to_shape(input, shape)


def mean(input: Tensor, axes: Optional[Union[int, list, tuple]] = None) -> Tensor:
    """Computes the mean of tensor elements over given axes.

    Args:
        input (Tensor): The input tensor.
        axes (Optional[Union[int, list, tuple]]): The axes to reduce.
    """
    return c_mean(input, axes)


def exp(input: Tensor) -> Tensor:
    """Computes the element-wise exponential of a tensor."""
    return c_exp(input)


def log(input: Tensor) -> Tensor:
    """Computes the element-wise natural logarithm of a tensor."""
    return c_log(input)


def log10(input: Tensor) -> Tensor:
    """Computes the element-wise base-10 logarithm of a tensor."""
    return c_log10(input)


def log2(input: Tensor) -> Tensor:
    """Computes the element-wise base-2 logarithm of a tensor."""
    return c_log2(input)


def sin(input: Tensor) -> Tensor:
    """Computes the element-wise sine of a tensor."""
    return c_sin(input)


def cos(input: Tensor) -> Tensor:
    """Computes the element-wise cosine of a tensor."""
    return c_cos(input)


def tanh(input: Tensor) -> Tensor:
    """Computes the element-wise hyperbolic tangent of a tensor."""
    return c_tanh(input)


def sinh(input: Tensor) -> Tensor:
    """Computes the element-wise hyperbolic sine of a tensor."""
    return c_sinh(input)


def cosh(input: Tensor) -> Tensor:
    """Computes the element-wise hyperbolic cosine of a tensor."""
    return c_cosh(input)


def abs(input: Tensor) -> Tensor:
    """Computes the element-wise absolute value of a tensor."""
    return c_abs(input)


def sign(input: Tensor) -> Tensor:
    """Computes the element-wise sign of a tensor (-1, 0, or 1)."""
    return c_sign(input)


def clamp(input: Tensor, min: float, max: float) -> Tensor:
    """Clamps all elements in a tensor into the range [min, max]."""
    return c_clamp(input, min, max)


__all__ = [
    "add",
    "sub",
    "mul",
    "div",
    "pow",
    "sqrt",
    "sum",
    "sum_to_shape",
    "mean",
    "exp",
    "log",
    "log10",
    "log2",
    "sin",
    "cos",
    "tanh",
    "sinh",
    "cosh",
    "abs",
    "sign",
    "clamp",
]
